File: ogn-db/src/ideas.rs
use diesel::PgConnection;
use diesel::prelude::*;

use crate::{Idea, IdeaId, ideas, NewIdea};
use crate::result::Result;
use crate::schema;

pub fn create_idea(conn: &mut PgConnection, label: &str) -> Result<Idea> {
    let new_idea = NewIdea {
        label: label.to_string(),
    };

    diesel::insert_into(schema::ideas::table)
        .values(new_idea)
        .get_result(conn)
        .map_err(|e| str_err!("Error saving new idea {}", e))
}

pub fn get_idea(conn: &mut PgConnection, id: IdeaId) -> Result<Option<Idea>> {
    schema::ideas::table
        .filter(schema::ideas::id.eq(id))
        .get_result(conn)
        .optional()
        .map_err(|e| str_err!("Error loading idea {}", e))
}

pub fn get_ideas(conn: &mut PgConnection, page_index: i64, page_size: i64) -> Result<Vec<Idea>> {
    schema::ideas::table
        .limit(page_size)
        .offset(page_index * page_size)
        .load::<Idea>(conn)
        .map_err(|e| str_err!("Error loading ideas {}", e))
}

pub fn idea_exists(conn: &mut PgConnection, id: IdeaId) -> Result<bool> {
    Ok(get_idea(conn, id)?.is_some())
}

pub fn get_num_ideas(conn: &mut PgConnection) -> Result<i64> {
    schema::ideas::table
        .select(diesel::dsl::count_star())
        .first(conn)
        .map_err(|e| str_err!("Error loading ideas {}", e))
}

pub fn delete_idea(conn: &mut PgConnection, idea_id: IdeaId) -> Result<usize> {
    diesel::delete(schema::ideas::table.filter(schema::ideas::id.eq(idea_id)))
        .execute(conn)
        .map_err(|e| str_err!("Error deleting idea {}", e))
}

File: OgnBackend/src/db/ideas.jl
@inline function create_idea(pool, idea::NewIdea)
    query = raw"INSERT INTO ideas (label) VALUES ($1) RETURNING id;"
    result = async_execute(pool, query, (idea.label,))

    @async begin
        result = result |> fetch
        cols = Tables.columns(result)

        if isempty(cols.id)
            nothing
        else
            Idea(IdeaId(cols.id |> first), idea.label)
        end
    end
end

@inline function get_idea(pool, idea_id::IdeaId)
    query = raw"SELECT * FROM ideas WHERE id = $1;"

    result = async_execute(pool, query, (idea_id,))

    @async begin
        result = result |> fetch
        cols = Tables.columns(result)

        if isempty(cols.id)
            nothing
        else
            @assert length(cols.id) == 1
            Idea(IdeaId(cols.id |> first), cols.label |> first)
        end
    end
end

@inline function get_ideas(pool)
    query = raw"SELECT * FROM ideas;"

    result = async_execute(pool, query)

    @async begin
        result = result |> fetch
        cols = Tables.columns(result)

        [Idea(IdeaId(id), label) for (id, label) in zip(cols.id, cols.label)]
    end
end

@inline function get_ideas(pool, page_num_size)
    page_num, page_size = page_num_size

    query = raw"SELECT * FROM ideas LIMIT $1 OFFSET $2;"

    result = async_execute(pool, query, (page_size, page_size * page_num))

    @async begin
        result = result |> fetch
        cols = Tables.columns(result)

        [Idea(IdeaId(id), label) for (id, label) in zip(cols.id, cols.label)]
    end
end

@inline function get_num_ideas(pool)
    query = raw"SELECT COUNT(*) FROM ideas;"

    result = async_execute(pool, query)

    @async begin
        result = result |> fetch
        cols = Tables.columns(result)

        if isempty(cols.count)
            nothing
        else
            @assert length(cols.count) == 1
            cols.count |> first
        end
    end
end

idea_exists(pool, idea_id::IdeaId) = @async begin
    result = get_idea(pool, idea_id) |> fetch
    result === nothing ? false : true
end

@inline function delete_idea(pool, idea_id::IdeaId)
    query = raw"DELETE FROM ideas WHERE id = $1;"
    result = async_execute(pool, query, (idea_id,))

    @async begin
        result = result |> fetch
        nothing
    end
end
